third_party/move/tools/move-decompiler/src/lib.rs

```diff
@@ -0,0 +1,208 @@
+pub struct Decompiler {
+    options: Options,
+    env: GlobalEnv,
+}
+
+impl Decompiler {
+    /// Creates a new decompiler.
+    pub fn new(options: Options) -> Self {
+        Self {
+            options,
+            env: GlobalEnv::new(),
+        }
+    }
+
+    /// Performs a run of the decompiler as specified by the options. This reads all
+    /// bytecode input files and runs the stages of decompilation for each, producing
+    /// output source files.
+    pub fn run<W>(mut self, error_writer: &mut W) -> anyhow::Result<()>
+    where
+        W: Write + WriteColor,
+    {
+        for input_path in mem::take(&mut self.options.inputs) {
+            let input_file = Path::new(Path::new(&input_path).file_name().unwrap_or_default())
+                .to_string_lossy()
+                .to_string();
+            let bytes =
+                fs::read(&input_path).with_context(|| format!("reading `{}`", input_path))?;
+            let output = if self.options.script {
+                let script = CompiledScript::deserialize(&bytes)
+                    .with_context(|| format!("deserializing script `{}`", input_path))?;
+                let source_map = self.get_source_map(&input_path, &input_file, &bytes);
+                self.decompile_script(script, source_map)
+            } else {
+                let module = CompiledModule::deserialize(&bytes)
+                    .with_context(|| format!("deserializing module `{}`", input_path))?;
+                let source_map = self.get_source_map(&input_path, &input_file, &bytes);
+                self.decompile_module(module, source_map)
+            };
+            self.env.check_diag(
+                error_writer,
+                Severity::Warning,
+                &format!("decompiling `{}`", input_path),
+            )?;
+            let out_file = Path::new(&self.options.output_dir)
+                .join(&input_file)
+                .with_extension(&self.options.ending);
+            fs::write(&out_file, output)
+                .with_context(|| format!("writing `{}`", out_file.display()))?;
+        }
+        Ok(())
+    }
+
+    fn get_source_map(&mut self, input_path: &str, input_file: &str, bytes: &[u8]) -> SourceMap {
+        let path = Path::new(&self.options.source_map_dir)
+            .join(input_file)
+            .with_extension("mvsm");
+        match fs::read(path) {
+            Ok(bytes) => bcs::from_bytes::<SourceMap>(&bytes)
+                .unwrap_or_else(|_| self.empty_source_map(input_path, &bytes)),
+            _ => self.empty_source_map(input_path, bytes),
+        }
+    }
+
+    /// Decompiles the given binary module. A source map must be provided for error
+    /// reporting; use `self.empty_source_map` to create one if none is available.
+    pub fn decompile_module(&mut self, module: CompiledModule, source_map: SourceMap) -> String {
+        // Load module into environment, creating stubs for any missing dependencies
+        let module_id = self
```

Comment:
Looks like the `load_compiled_module` expects that the `module` is already verified. Are we expected to do an explicit verification here (at the call site)?